python/cudf/cudf/core/window/rolling.py

```diff
@@ -319,17 +328,73 @@ def _normalize(self):
                 raise ValueError(
                     "window must be integer or " "convertible to a timedelta"
                 ) from e
-
-            window = cudautils.window_sizes_from_offset(
-                self.obj.index._values.data_array_view, window
-            )
             if self.min_periods is None:
                 min_periods = 1
 
-        self.window = window
+        self.window = self._window_to_window_sizes(window)
         self.min_periods = min_periods
 
+    def _window_to_window_sizes(self, window):
+        """
+        For non-fixed width windows,
+        convert the window argument into window sizes.
+        """
+        if pd.api.types.is_integer(window):
+            return window
+        else:
+            return cudautils.window_sizes_from_offset(
+                self.obj.index._values.data_array_view, window
+            )
+
     def __repr__(self):
         return "{} [window={},min_periods={},center={}]".format(
             self.__class__.__name__, self.window, self.min_periods, self.center
         )
+
+
+class RollingGroupby(Rolling):
+    def __init__(self, groupby, window, min_periods=None, center=False):
+        """
+        Grouped rolling window calculation.
+
+        See also
+        --------
+        cudf.core.window.Rolling
+        """
+        sort_order = groupby.grouping.keys.argsort()
+        self._group_keys = groupby.grouping.keys.take(sort_order)
+        obj = groupby.obj.take(sort_order)
```

Comment:
Are the keys included in `obj` as well? If so you're unnecessarily double gathering them.

Reply:
They may not be, as you can group an object by external columns. :(

Reply:
Do you think the common case is that they'll be in `groupby.obj`? Would it be worth traversing through the columns in `obj` to see if the column(s) in `groupby.grouping.keys` are within `obj` to do a single gather?

Reply:
> Do you think the common case is that they'll be in `groupby.obj`

I don't know - I'd think the external groupby is common. We have to account for a combination of internal and external columns in `keys`.

> Would it be worth traversing through the columns in obj to see if the column(s) in groupby.grouping.keys are within obj to do a single gather?

It would mean either:

1. Popping items out from `obj`, doing the gather on the remaining columns, and then inserting gathered columns from `keys` back into the right locations of `obj`. 

2. Popping items out from `keys`, doing the gather on the remaining columns, and then inserting gathered columns from `obj` back into the right locations of `keys`.

My immediate feeling is that the simplicity of the current approach should be preferred. But if you feel strongly, happy to look into it :)

Reply:
Maybe lets just add a `TODO` comment that there's a possible optimization to do the gather in one shot.

Reply:
Done